### PR TITLE
Implement global and per-ip connection rate throttles

### PR DESCRIPTION
**Pull request by Vagabond**
Essentially allow grpcbox to control how quickly connections will be opened. Generally the global connection throttle should be set higher than the per-ip one, the per-ip one should be low so that spamming clients can quickly be shut down while the global one should be higher because all non ip-limited connections will count towards the global limit.

### CODE REVIEW

File: src/grpcbox_acceptor.erl [hunk 1]
```diff
@@ -40,8 +42,25 @@ acceptor_terminate(Reason, _) ->
...
+    case throttle:check(PoolName, PeerName) of
+        {ok, _, _} ->
+            check_any_allowed(PoolName);
+        rate_not_set ->
+            check_any_allowed(PoolName);
+        {limit_exceeded, _, _} -> false
+    end.
+
+check_any_allowed(PoolName) ->
+    case throttle:check({PoolName, all}, all) of
```
Comment (andymck) [inline src/grpcbox_acceptor.erl:61]: what is 'all' representing here in the scope ? on the setup you are specifying `any`

File: src/grpcbox_services_sup.erl [hunk 2]
```diff
@@ -48,6 +48,23 @@ init([ServerOpts, GrpcOpts, ListenOpts, PoolOpts, TransportOpts, ServiceSupName]
 
     %% unique name for pool based on the ip and port it will listen on
     PoolName = pool_name(ListenOpts),
+    case {maps:get(accept_rate, ListenOpts, undefined),
```
Comment (andymck) [inline src/grpcbox_services_sup.erl:51]: The max_connection settings is specified in the server opts, whilst the rate limiting here is in the listen opts.  Should we keep them in the same section ?
The listenopts are not threaded through as part of the acceptor state, hence why the max_connections is in server opts.  It also keeps it with the existing `max_concurrent_streams` settings

File: src/grpcbox_acceptor.erl (hunk 1 above)
Comment (andymck) [reply src/grpcbox_acceptor.erl]: confirmed it was a type...defaulted to using `any`

File: src/grpcbox_acceptor.erl [hunk 3]
```diff
@@ -16,7 +16,8 @@ acceptor_continue(_PeerName, Socket, {ssl, _MRef, PoolName, ServerOpts, Chatterb
     case ssl:negotiated_protocol(AcceptSocket) of
         {ok, <<"h2">>} ->
             MaxConns = maps:get(max_connections, ServerOpts, unlimited),
-            case connection_allowed(PoolName, MaxConns) of
+            {ok, {PeerName, _PeerPort}} = ssl:peername(AcceptSocket),
```
Comment (jeffgrunewald) [inline src/grpcbox_acceptor.erl:19]: we don't use ssl for grpc connections, is that going to be a problem here? should we use the `socket:peername/1` instead?

File: src/grpcbox_services_sup.erl (hunk 2 above)
Comment (andymck) [reply src/grpcbox_services_sup.erl]: moved all to ServerOpts

File: src/grpcbox_acceptor.erl (hunk 3 above)
Comment (andymck) [reply src/grpcbox_acceptor.erl:19]: there is a diff handler for tcp connection

**Review by jeffgrunewald — APPROVED** [review]

**Pull request merged**